### PR TITLE
Editor RAI tab

### PR DESCRIPTION
**Pull request by JoanGi**
As a subsequent PR of the [library adaptation for RAI attributes](https://github.com/mlcommons/croissant/pull/553), this PR adds the RAI tab to the editor. 

The tab follows the specification in http://mlcommons.org/croissant/RAI/ (near to be published)

Captura de pantalla 2024-02-28 a las 12 41 37

There are also a few changes in the name of the serialize/deserialize function of the Python library.

After this PR, there is planned a PR to update some example datasets with RAI attributes

### CODE REVIEW

**Comment by JoanGi** [comment]
I've noticed that 1-to-many RAI attributes were not working correctly if there was only 1 element.  @marcenacp solved it, by implementing the unbox_singletion_list() for other Croissant 1-to-many attributes while serializing Metadata to JSON. It works like a charm, so[ I've applied](https://github.com/mlcommons/croissant/pull/578/commits/d60c574638203babb972d195b8567902de3781eb) it to the RAI attributes that are 1-to-many during the serialization. Now, the editor generated the examples correctly.

**Comment by JoanGi** [comment]
The editor is now adapted to the 1.0 RAI spec naming. Does that sound OK to you @marcenacp and @benjelloun?

**Comment by benjelloun** [comment]
Thanks for pushing through this! Out of caution, can we hold off just a few days, just to make sure we don't break anything in the editor during the Croissant launch?

**Comment by JoanGi** [comment]
Sure! Let's merge it the following Monday, then. :)

**Comment by JoanGi** [comment]
Do you think we can proceed, or is it better to still wait a few days?

**Comment by marcenacp** [comment]
> Do you think we can proceed, or is it better to still wait a few days?

I think we can proceed. Can you please rebase before merging to resolve conflicts?

**Comment by JoanGi** [comment]
Rebased with the latest library updates and fix some naming changes in the library's RAI properties.

**Comment by JoanGi** [comment]
@marcenacp, does the RAI tab look ok for you?

File: editor/views/metadata.py
Comment (marcenacp) [inline editor/views/metadata.py:18]: Remove this whole column if it doesn't provide any information?

File: editor/views/rai.py
Comment (marcenacp) [inline editor/views/rai.py:70]: Remove the trailing space in the label (also later in the file)

File: editor/views/rai.py
Comment (marcenacp) [inline editor/views/rai.py:147]: It seems you reuse the same "component" several times. Can you factorize?

File: editor/core/state.py
Comment (marcenacp) [inline editor/core/state.py:216]: Remove?

**Review by marcenacp — CHANGES_REQUESTED** [review]
@JoanGi Sorry I was OOO for a few days, I just commented on the PR. Thanks!

File: editor/views/rai.py
Comment (marcenacp) [reply editor/views/rai.py:121]: The component doesn't exist yet, but you could create one.

**Comment by marcenacp** [comment]
Maybe later on we should also write end-to-end tests for the RAI tab - at least to test list-based fields with append/remove. The e2e tests are in the [`cypress/` folder](https://github.com/mlcommons/croissant/tree/main/editor/cypress/e2e). What do you think?

File: editor/views/rai.py
Comment (marcenacp) [reply editor/views/rai.py]: Thanks!

File: editor/events/rai.py
Comment (JoanGi) [reply editor/events/rai.py]: Yes, much cleaner. Now handle_rai_change gets index as input. Thank you!

File: editor/views/rai.py
Comment (JoanGi) [reply editor/views/rai.py:121]: Now I get it! 

I've created a component to handle 1toMany Attributes. Reduced the size of the file by half. Thank you for the recommendation!

File: editor/views/rai.py
Comment (marcenacp) [inline editor/views/rai.py:122]: In Python, we recommend using snake_case over camelCase

File: editor/views/rai.py
Comment (marcenacp) [inline editor/views/rai.py:163]: I would remove all the comments as the title is explicit enough

File: editor/views/rai.py
Comment (marcenacp) [inline editor/views/rai.py:215]: `title`? (Or use snake_case)

File: editor/views/rai.py
Comment (marcenacp) [inline editor/views/rai.py:219]: At the beginning, you can have:

```
if not isinstance(attribute, list):
  attribute = [attribute]
```

and then all the logic should be the same and you spare if/else.

But why don't we only pass lists in the first place?

File: editor/views/rai.py
Comment (marcenacp) [inline editor/views/rai.py:165]: With many arguments, it's better to specify them explicitly for readability:

```
oneToManyProperty(
  title="**Data biases**",
  ...
```

File: editor/views/rai.py
Comment (marcenacp) [inline editor/views/rai.py:216]: Can you add a docstring to specify the behaviour? Especially for `attribute`, it's tricky to understand when it should be a list or not (and why)

File: editor/views/rai.py
Comment (marcenacp) [inline editor/views/rai.py:220]: singleAttribute->single_attribute

Should attribute be named attributes (to reflect that it's a list)?

**Review by marcenacp — COMMENTED** [review]
Thanks! I love the new component. Still a few nits

File: editor/views/rai.py
Comment (JoanGi) [reply editor/views/rai.py]: Changed.

File: editor/views/rai.py
Comment (JoanGi) [reply editor/views/rai.py]: removed

File: editor/views/rai.py
Comment (JoanGi) [reply editor/views/rai.py]: There should be cases where the input is not a list. For instance, if we load a Croissant file in the editor with any 1toN attribute not declared as a list. In that sense, it seems more secure to me to handle the casting logic of "list" inside the component. I've factorized the component following your recommendation and checking for a list at the beginning, much simpler and readable. :)

File: editor/views/rai.py
Comment (JoanGi) [reply editor/views/rai.py]: Now calls have explicitly declared params.

File: editor/views/rai.py
Comment (JoanGi) [reply editor/views/rai.py]: fixed

File: editor/views/rai.py
Comment (JoanGi) [reply editor/views/rai.py:212]: added

File: editor/views/rai.py
Comment (JoanGi) [reply editor/views/rai.py]: changed to title

**Comment by JoanGi** [comment]
Looks good to you, @marcenacp?

File: editor/events/rai.py
Comment (marcenacp) [inline editor/events/rai.py:37]: Remove?

**Comment by benjelloun** [comment]
Small UI request: Can you move the RAI tab after the resources and recordsets one? We should also leave a bit of space between the main tabs and the extension(s) one(s).

Thanks!
Omar

File: editor/views/rai.py
Comment (marcenacp) [inline editor/views/rai.py:240]: Prefer callbacks rather than st.rerun (see https://docs.streamlit.io/develop/api-reference/execution-flow/st.rerun). Maybe by having an `on_add`/`on_remove`? That'd allow to distinguish between the case list/not list.

```python
one_to_many_property(
  attributes=...,
  handle_add=...,
  handle_remove=...,
)


if handle_add:
  st.button("+ add", key=f"add_{key}", on_click=handle_add)
```

**Review by marcenacp — APPROVED** [review]
Hey Joan, LGTM, but I think we could still improve the one_to_many_property function. Thanks!

**Pull request merged**
